Don't run docker containers as root

jeremyestein opened the pull request:
Fixes #234. Waiting for testing on GAE before merging.

Note the addition of two new variables `PIXL_USER_UID` and `PIXL_USER_GID`

Firstly merge all the Dockerfiles for images that we control (imaging, export, hasher) to make this process easier.

Run all our python containers as the user/group pixl, which we create as part of the build process, using the UID/GID as specified in the config.

Export API mounts export dir read-only as it doesn't need to write any more.

Document how the host must be set up for this to work.

Do same for orthanc images.

jeremyestein committed:
Make packages easier to read

jeremyestein committed:
Remove unneeded packages

jeremyestein committed:
Merge three pixl python docker images into one multi-stage dockerfile to
avoid repetition.

jeremyestein committed:
Specify healthcheck command in only one place

jeremyestein committed:
De-dupe the pre-reqs installation code using a build arg

jeremyestein committed:
De-dupe the actual install as well

jeremyestein committed:
Add (failing) test for the condition we are aiming for: containers to
run as non-root users.

jeremyestein committed:
Run as user pixl, which we have to create inside the image with a
fixed user ID.

jeremyestein committed:
The system test already builds the required docker images, and passes
the correct env files in. No need to do it here as well, and it was
failing due to lack of build args.

jeremyestein committed:
Semi-WIP: Create pixl user+group on the GHA test runner so that we can set file
permissions correctly for the export directory, which is needed so that
it can be deleted from inside the container.

jeremyestein committed:
Fix variable usage and exports dir location

jeremyestein committed:
More debugging

jeremyestein committed:
debug

jeremyestein committed:
debug

jeremyestein committed:
split build and up

jeremyestein committed:
Create user home dir in the GHA script to avoid error from failed
creation later on

jeremyestein committed:
Why is pytest not found? Some kind of env change caused by sudo?

jeremyestein committed:
Preserve path when running sudo

jeremyestein committed:
Run tests as normal runner user

jeremyestein committed:
Don't see why this needs to have docker group

jeremyestein committed:
Debugging for perm failure on file deletes

jeremyestein committed:
Fix scope error

jeremyestein committed:
OK maybe this was needed after all

jeremyestein committed:
Since CLI creates everything in exports dir these days, it makes more
sense to delete stuff afterwards from the host side rather than the
container side.

jeremyestein committed:
Try not creating home dir for pixl - what is even using this?

jeremyestein committed:
and therefore I can't do this any more

jeremyestein committed:
Remove debugging

jeremyestein committed:
Document the reason for the permissions setup better. Don't need to
export variables any more.

jeremyestein committed:
If Export API is believed to only read from the export dir, then let's
enforce this to avoid any weird surprises.

jeremyestein committed:
First attempt at documentation on docker permissions.

jeremyestein committed:
CLI user needs to be able to write to export dir.

jeremyestein committed:
Clarify docs

jeremyestein commented:
In writing the docs, I have come up with what I think is a problem, that will likely need a user/group adding to the GAE to fix. Let me quote from the docs:

---

Imagine you are a user on the GAE with username `fred`, primary group `fred`, and supplementary
group `docker`.

When you run the CLI, it will run as `fred:fred`, and normally any files it creates would also have
this ownership. (ACLs defined in `/gae` will affect this however, so the files may have
a different group ownership, such as `fred:docker`)

There is a problem here in that the export API will likely be running as some other user/group,
so it will not be able to read the files.

Options for fixing:
* The CLI (or an ACL on the exports dir) makes the files world readable. Not ideal from a security
  pov.
* Relying on the existing ACL for `/gae` which gives files group ownership of `docker`, we run the
   Export API as the `docker` group, thus giving it read access to these files. However, this is
   very similar to running the container as root, which is what this whole thing is trying to avoid.
* Create a new group (`pixl`) on the GAE and add all pixl devs to it.
  Set an ACL on the exports dir that sets group ownership for all created files as `pixl`.
  Then we could set its GID in the env file so that the container runs as group `pixl`.
* Always build and run the containers as the same user that is running the CLI. You set the
  build vars `PIXL_USER_[UG]ID` to the numerical value of fred:fred. PIXL runs as you. When you run
  CLI, the files are owned by fred:docker, so export api can read them fine. This lacks flexibility as we should
  all be able to run PIXL.

peshence commented on docs/design/docker_perms.md:
```diff
@@ -0,0 +1,83 @@
+So, all files in the export temp dir need to be readable by the Export API container, and therefore
+it matters as what user/group the CLI is run.
+
+### Example
+
+Imagine you are a user on the GAE with username `fred`, primary group `fred`, and supplementary
+group `docker`.
+
+When you run the CLI, it will run as fred:fred, and normally any files it creates would also have
+this ownership. (ACLs defined in `/gae` will affect this however, so the files may have
+a different group ownership, such as fred:docker)
+
+There is a problem here in that the export API will likely be running as some other user/group,
+so it will not be able to read the files.
+
+Options for fixing:
+* The CLI (or an ACL on the exports dir) makes the files world readable. Not ideal from a security
+  pov.
+* Relying on the existing ACL for `/gae` which gives files group ownership of `docker`, we run the
+   Export API as the `docker` group, thus giving it read access to these files. However, this is
+   very similar to running the container as root, which is what this whole thing is trying to avoid.
+* Create a new group (`pixl`) on the GAE and add all pixl devs to it.
+  Set an ACL on the exports dir that sets group ownership for all created files as `pixl`.
+  Then we could set its GID in the env file so that the container runs as group `pixl`.
+* Always build and run the containers as the same user that is running the CLI. You set the
+  build vars `PIXL_USER_[UG]ID` to the numerical value of fred:fred. PIXL runs as you. When you run
+  CLI, the files are owned by fred:docker, so export api can read them fine.
```
I also just wrote this on Slack, but for posterity:

> I assume that we intend to have the apis running in the background indefinitely, started by a specific role user and that multiple different users may start the pipeline from the CLI, so the new group approach makes the most sense to me.
> I also like the idea of the CLI not actually containing functionality and only accessing APIs, but would say that is a separate issue/PR

This portion of the docs would then be converted to just instructions about having the group, do I understand correctly?

jeremyestein replied:
Yes, it would be converted to instructions as soon as I work out what those instructions are. Then this bit could perhaps go into the body of an issue that recommends refactoring the CLI.

peshence commented on test/conftest.py:
```diff
@@ -135,43 +134,19 @@ def _upload_dicom_instance(dicom_dir: Path, **kwargs: Any) -> None:
     run_subprocess(["pixl", "populate", str(RESOURCES_OMOP_DIR.absolute())], TEST_DIR)
     # poll here for two minutes to check for imaging to be processed, printing progress
     wait_for_stable_orthanc_anon(121, 5, 15, min_instances=3)
-    yield
-    run_subprocess(
-        [
-            "docker",
-            "exec",
-            "system-test-export-api-1",
-            "rm",
-            "-r",
-            "/run/projects/exports/test-extract-uclh-omop-cdm/",
-        ],
-        TEST_DIR,
-    )
```
🚀

peshence commented on .github/workflows/main.yml:
```diff
@@ -128,6 +119,18 @@ jobs:
               run: |
+                  # We need PIXL_USER_UID and PIXL_USER_GID from the test env file.
+                  source .env
+                  # Export API (running as pixl) needs to be able to *read* files in
+                  # the export dir, but it doesn't write there.
+                  # So, set ownership/permissions accordingly.
+                  # Bear in mind that ownership/perms of files created below this dir will
+                  # depend on the CLI process. I think we don't need to set ACLs for now.
```
```suggestion
                  # Export API (running as pixl) needs only *read* files in the export dir.
                  # Ownership/perms of files created in the export dir will
                  # depend on the user running the CLI process.
```

Not adamant about the wording but would remove the opinions/questions and make it shorter and more to the point

peshence reviewed (approved):
Looks good

jeremyestein committed:
Go with the assumption that the host machine will have a "pixl" user and
group which we run our containers as, and human users will be added to the
pixl group.

jeremyestein committed:
oops left in empty arg

jeremyestein committed:
Set the setgid bit so that subdirs/files in exports will be owned by the
pixl group

jeremyestein committed:
debugging

jeremyestein committed:
fix debugging

jeremyestein committed:
doesn't exist?

jeremyestein committed:
Make debugging more useful and permanent. More docs

jeremyestein committed:
more debugging

jeremyestein committed:
Set the FACL

jeremyestein committed:
usermod doesn't change groups of existing shells. Need to run the system
test in a new shell.

jeremyestein committed:
Invoking a new shell with "bash" doesn't re-read groups, but using sudo
does.

jeremyestein commented:
This change can't go in until the `pixl` user and group have been created on the GAE, and the pixl dev users added to the group.

jeremyestein committed:
Need to preserve envs

jeremyestein committed:
more debugging

jeremyestein committed:
Unlikely to work because su will prompt

jeremyestein committed:
Will just passing PATH be enough?

jeremyestein committed:
Pass through all env as well as PATH

jeremyestein committed:
Remove debugging

dram1964 commented:
> This change can't go in until the `pixl` user and group have been created on the GAE, and the pixl dev users added to the group.

Hi @jeremyestein, I've added PIXL user and group to GAE05 - and added you to the group. Let me know how that works for you.

dram1964 reviewed (approved):
Using a 'service account' for this sort of thing seems to be a good idea

jeremyestein committed:
Merge branch 'main' into jeremy/docker-uid-new

jeremyestein committed:
Merge branch 'main' into jeremy/docker-uid-new

jeremyestein committed:
Unite orthanc Dockerfiles

jeremyestein committed:
Add orthanc containers to test

jeremyestein committed:
Make orthanc run as orthanc user instead of root

jeremyestein committed:
Can also use non-root for fakes in system test

jeremyestein committed:
Non-root orthanc works for core and imaging tests too

jeremyestein committed:
Add fakes to system test container user check

jeremyestein committed:
Remove debugging

jeremyestein committed:
Use same orthanc version everywhere

jeremyestein committed:
Use existing YAML anchors

jeremyestein committed:
Don't think this is needed since we moved this code to the export-api

stefpiatek committed:
Try adding orthanc to PIXL group

stefpiatek committed:
Try adding orthanc to PIXL group

stefpiatek committed:
Try adding orthanc to PIXL group

stefpiatek committed:
Try adding orthanc to PIXL group

stefpiatek committed:
Change owner of pixl file

stefpiatek committed:
Try setting pixl user before copying

stefpiatek committed:
Change owner of python projects on copy

stefpiatek committed:
Change owner of python projects on copy

stefpiatek committed:
Change owner of python projects on copy

stefpiatek committed:
Merge branch 'main' into jeremy/docker-uid-new

stefpiatek committed:
Define wait for images to be exported in conftest

stefpiatek committed:
Add group ID to test build args

milanmlft committed:
Merge branch 'main' into jeremy/docker-uid-new

milanmlft commented on README.md:
````diff
@@ -295,10 +295,14 @@ These files will subsequently be uploaded to the `parquet` destination specified
 ....................................................../omop/public/*.parquet
 ```
 
+This directory is current located in `projects/exports` relative to the source code root.
````
```suggestion
This directory is currently located in `projects/exports` relative to the source code root.
```

milanmlft commented on docker/hasher-api/Dockerfile:
❤️

stefpiatek committed:
Update README.md

Co-authored-by: Milan Malfait <m.malfait@ucl.ac.uk>

milanmlft committed:
Merge branch 'main' into jeremy/docker-uid-new

milanmlft committed:
Consolidate `RUN` commands

Following good practice:
https://github.com/hadolint/hadolint/wiki/DL3059

stefpiatek committed:
Merge branch 'main' into jeremy/docker-uid-new

milanmlft committed:
Add @dram1964's instructions for setting up a PIXL user on GAE

stefpiatek committed:
Merge remote-tracking branch 'origin/main' into jeremy/docker-uid-new

# Conflicts:
#	.github/workflows/main.yml
#	docker/hasher-api/Dockerfile
#	docker/imaging-api/Dockerfile
#	docker/orthanc-anon/Dockerfile
#	docker/orthanc/Dockerfile
#	docker/pixl-python/Dockerfile
#	pixl_imaging/tests/docker-compose.yml
#	test/conftest.py

stefpiatek committed:
Merge remote-tracking branch 'origin/main' into jeremy/docker-uid-new

# Conflicts:
#	.github/workflows/main.yml
#	docker/hasher-api/Dockerfile
#	docker/imaging-api/Dockerfile
#	docker/orthanc-anon/Dockerfile
#	docker/orthanc/Dockerfile
#	docker/pixl-python/Dockerfile
#	pixl_imaging/tests/docker-compose.yml
#	test/conftest.py

jeremyestein committed:
Merge branch 'main' into jeremy/docker-uid-new

jeremyestein commented on docker/orthanc/Dockerfile:
```diff
@@ -37,17 +40,48 @@ COPY ./pixl_dcmd/ /pixl_dcmd
+COPY --chown=orthanc:orthanc ./orthanc/orthanc-anon/plugin/pixl.py /etc/orthanc/pixl.py
+COPY --chown=orthanc:orthanc ./orthanc/orthanc-anon/config /run/secrets
+
 RUN sed -i "s/\${ORTHANC_CONCURRENT_JOBS}/${ORTHANC_CONCURRENT_JOBS:-5}/g" /run/secrets/orthanc.json
```
`ORTHANC_CONCURRENT_JOBS` doesn't seem to be declared as a build ARG for `pixl_orthanc_anon`. Is it just reverting to the given default value?

jeremyestein commented:
Have split the refactoring side of this into #583 , which could possibly be merged with less risk than this PR. If we do decide to merge this, I think the fixup commit in the other branch (98fa6896fb33bf6090fb5b9c2309dbe64ce331bd) should be applied here too.

stefpiatek commented:
I expect this is failing because the `PIXL_DB_*` env variables were removed from anon?
